bolt/kv.go

```diff
@@ -233,26 +235,49 @@ func (s *KVStore) Restore(ctx context.Context, r io.Reader) error {
 
 		// Swap and reopen under lock.
 		s.mu.Lock()
-		defer s.mu.Unlock()
-
-		if err := s.db.Close(); err != nil {
+		if err := s.swapAndOpenRestored(); err != nil {
+			s.mu.Unlock()
 			return err
 		}
 
-		// Atomically swap temporary file with current DB file.
-		if err := fs.RenameFileWithReplacement(s.tempPath(), s.path); err != nil {
+		// Now that the DB has been re-opened, release the lock so that the
+		// migrations can be run.
+		s.mu.Unlock()
+
+		// Run migrations on the restored database. This will catch the restored
+		// database up to the current state expected by the application.
+		restoreMigrator, err := migration.NewMigrator(
+			s.log.With(zap.String("service", "bolt restore migrations")),
+			s,
+			all.Migrations[:]...,
+		)
+		if err != nil {
 			return err
 		}
 
-		// Reopen with new database file.
-		return s.openDB()
+		return restoreMigrator.Up(ctx)
 	}(); err != nil {
 		os.Remove(s.tempPath()) // clean up on error
 		return err
 	}
 	return nil
 }
 
+// swapAndOpenRestored is used while restoring a database to close the existing
+// database, replace the existing database with the restored database at the
+// temp path, and then re-open the restored database.
+func (s *KVStore) swapAndOpenRestored() error {
```

Comment:
This function was re-factored from existing code into a separate function to make it cleaner to manage the mutex locking and unlocking.